cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -45,6 +46,43 @@ public WorkflowSpecification(String className, String name, String description,
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<>(properties));
     this.nodes = Collections.unmodifiableList(new ArrayList<>(nodes));
+    this.nodeIdMap = Collections.unmodifiableMap(generateNodeIdMap());
+  }
+
+  /**
+   * Visit all the nodes in the {@link Workflow} and generate the map of node id to
+   * the {@link WorkflowNode}. Returned map does not contain any mapping for the FORK node,
+   * however contains the ACTION and CONDITION nodes within the FORK node itself.
+   */
+  private Map<String, WorkflowNode> generateNodeIdMap() {
+    Map<String, WorkflowNode> nodeIdMap = new HashMap<>();
+    Queue<WorkflowNode> nodes = new LinkedList<>(this.nodes);
+    while (!nodes.isEmpty()) {
+      WorkflowNode node = nodes.poll();
+      switch (node.getType()) {
+        case ACTION: {
+          nodeIdMap.put(node.getNodeId(), node);
+          break;
+        }
+        case FORK: {
+          WorkflowForkNode forkNode = (WorkflowForkNode) node;
```

Comment:
Why not add the fork node to the map as well? I think it's better for the list and map contain the same set of nodes.